services/gundeck/src/Gundeck/API.hs

```diff
@@ -131,7 +137,96 @@ sitemap = do
+docsH (url ::: _) =
   let doc = mkSwaggerApi (decodeLatin1 url) Model.gundeckModels sitemap
    in return $ json doc
+
+addTokenH :: UserId ::: ConnId ::: JsonRequest PushToken ::: JSON -> Gundeck Response
+addTokenH (uid ::: cid ::: req ::: _) = do
+  newtok <- fromJsonBody req
+  handleAddTokenResponse <$> Push.addToken uid cid newtok
+
+handleAddTokenResponse :: Push.AddTokenResponse -> Response
+handleAddTokenResponse = \case
+  Push.AddTokenSuccess newtok -> success newtok
+  Push.AddTokenNoBudget -> snsThreadBudgetReached
+  Push.AddTokenNotFound -> notFound
+  Push.AddTokenInvalid -> invalidToken
+  Push.AddTokenTooLong -> tokenTooLong
+  Push.AddTokenMetadataTooLong -> metadataTooLong
+
+success :: PushToken -> Response
+success t =
+  let loc = Text.encodeUtf8 . tokenText $ t ^. token
+   in json t & setStatus status201 & addHeader hLocation loc
+
+invalidToken :: Response
+invalidToken =
+  json (Error status400 "invalid-token" "Invalid push token")
+    & setStatus status404
+
+snsThreadBudgetReached :: Response
+snsThreadBudgetReached =
+  json (Error status400 "sns-thread-budget-reached" "Too many concurrent calls to SNS; is SNS down?")
+    & setStatus status413
+
+tokenTooLong :: Response
+tokenTooLong =
+  json (Error status400 "token-too-long" "Push token length must be < 8192 for GCM or 400 for APNS")
+    & setStatus status413
+
+metadataTooLong :: Response
+metadataTooLong =
+  json (Error status400 "metadata-too-long" "Tried to add token to endpoint resulting in metadata length > 2048")
+    & setStatus status413
+
+notFound :: Response
+notFound = empty & setStatus status404
+
+deleteTokenH :: UserId ::: Token ::: JSON -> Gundeck Response
+deleteTokenH (uid ::: tok ::: _) = setStatus status204 empty <$ Push.deleteToken uid tok
+
+listTokensH :: UserId ::: JSON -> Gundeck Response
+listTokensH (uid ::: _) = setStatus status200 . json <$> Push.listTokens uid
+
+pushH :: Request ::: JSON -> Gundeck Response
+pushH (req ::: _) = do
+  ps <- fromJsonBody (JsonRequest req)
+  empty <$ Push.push ps
+
+paginateH :: JSON ::: UserId ::: Maybe ByteString ::: Maybe ClientId ::: Range 100 10000 Int32 -> Gundeck Response
+paginateH (_ ::: uid ::: sinceRaw ::: clt ::: size) = do
+  Notification.PaginateResult gap page <- Notification.paginate uid (join since) clt size
+  pure . updStatus gap . json $ page
+  where
+    since :: Maybe (Maybe NotificationId)
+    since = parseUUID <$> sinceRaw
+    --
+    parseUUID :: ByteString -> Maybe NotificationId
+    parseUUID = UUID.fromASCIIBytes >=> isV1UUID >=> return . Id
+    --
+    isV1UUID :: UUID -> Maybe UUID
+    isV1UUID u = if UUID.version u == 1 then Just u else Nothing
+    --
+    -- FUTUREWORK: updStatus sets the status code to 404 in even if it returns notifications.
+    -- even if this is not a mistake, it certainly should be changed into something less
+    -- surprising.  (sync with clients before we change this!)
+    updStatus :: Bool -> Response -> Response
```

Comment:
This is business logic, and business logic with `Bool -> Response -> Response` is very blind.  Can we bubble this out of the turning things into `json` ?

Reply:
I had a similar comment, but apparently linked to that particular commit, so it doesn't show up in the files view.

> I'm not sure about this part, since now you have logic here, in the wrapper function. The typed paginate function should look at `since` and make decisions.

If we can change the behaviour to throw a 400 when `since` doesn't parse as V1 UUID, IMO that can be handled in the wrapper, but not the distinction between `since` being provided or not.